### DISCUSSION TITLE
CI Friendly versions and build of milestones and releases

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-tycho/tycho#2261] [General]
Hi, is it possible to use the CI Friendly versions feature to build maven milestones and release
(e.g. maven version 1.0.0 or 1.0.0.M1)
and still have valid qualifers in osgi bundles
(1.0.0.vtimestamp)

using the ci friendly versions feature as in tycho-its/projects/ci-friendly/buildqualifier
does not seem to work. it will produce invalid manifest

org.osgi.framework.BundleException: Invalid manifest header Bundle-Version: "1.0.0.M1.202303201403" : invalid version "1.0.0.M1.202303201403": invalid qualifier "M1.202303201403"
at org.eclipse.osgi.internal.resolver.StateBuilder.createBundleDescription (StateBuilder.java:124)

see also https://github.com/xtext/xtext-monorepo/issues/15
https://github.com/eclipse/xtext/issues/2135#issuecomment-1476054632

**Comment by laeubi** [comment]
> using the ci friendly versions feature as in tycho-its/projects/ci-friendly/buildqualifier
does not seem to work. it will produce invalid manifest

How do you call it to produce the invalid version? Your description is a bit confusing, because if you mention to produce a M1 version but wanted to have `v<timestamp>`, while the `v<timestamp>` is one way to format the qualifier it is not the only "valid" one.

**Reply by cdietrich** [reply]
https://github.com/cdietrich/tycho-ci-friendly-bq
call it with

`mvn clean install -DreleaseVersion=1.0.0 -Dqualifier=".M1"`
same with
`mvn clean install -DreleaseVersion=1.0.0.M1 -Dqualifier=""`

the goal is to have a normal p2 versioning with x.x.x.qualifier independent if the maven version is a release, milestone or snapshot

**Reply by laeubi** [reply]
The current supported approach is that you call it like described here:

https://tycho.eclipseprojects.io/doc/master/TychoCiFriendly.html

So your maven artifact + bundle get the same version what is the formatted version qualifier.

**Reply by cdietrich** [reply]
ok so we will keep bash scripts that do search and replace / a combination of maven and tycho versions plugin.

**Reply by laeubi** [reply]
@cdietrich it was not used before in this way, but what I can think about is that one simply uses not qualifier but some other variable (e.g. `${changelist}` or we can even define a new one like `${mavenQualifier}`), then it should give the desired results.

**Reply by laeubi** [reply]
By the way it would be good if you can submit your example as an [integration-test](https://github.com/eclipse/tycho/wiki#providing-an-integration-test) to demonstrate the issue.

**Reply by laeubi** [reply]
@cdietrich just wondering should a qualifer for Milestone not be 1.0.0-M1?

**Reply by cdietrich** [reply]
good question. have seen both variants in different projects. we currently do .m1

**Reply by laeubi** [reply]
At least maven treats everything unknown as a simple string but `-alpha` `-milestone` `-M` ... have special meanings

**Reply by laeubi** [reply]
Just for your information I have now checked with your example and actually there are two cases:

1. You specify the `-Dqualifier` then it will be used as is for maven part so I think you not need to use a different variable as it is only maven model
2. You **not** specify `-Dqualifier` but `-Dtycho.buildqualifier.format=yyyyMMddHHmm` then the qualifier maven model variable will get the formatted value of the timestamp provider.

What already would work is performing a release like that:

`mvn clean install -DreleaseVersion=1.0.0 -Dqualifier=`

this will result in a Bundle version of 1.0.0 and and a pom version of 1.0.0.

What not work (and I tried different combinations of this) is `mvn clean install -DreleaseVersion=1.0.0 -Dqualifier=-M1` (or similar), because in that case tycho combines the timestamped value with the maven version and only detects `-SNAPSHOT` as a special meaning.

I'm currently investigate how to circumvent the problem, what is quite tricky as maven even has the format `MAJOR.MINOR.MICRO.BUILD-QUALIFIER`
